src/sync.h

```diff
@@ -254,7 +254,22 @@ using DebugLock = UniqueLock<typename std::remove_reference<typename std::remove
 //!
 //!   int val = WITH_LOCK(cs, return shared_val);
 //!
-#define WITH_LOCK(cs, code) [&] { LOCK(cs); code; }()
+//! Note:
+//!
+//! Since the return type deduction follows that of decltype(auto), while the
+//! deducted type of:
+//!
+//!   WITH_LOCK(cs, return {int i = 1; return i;});
+//!
+//! is int, the deducted type of:
+//!
+//!   WITH_LOCK(cs, return {int j = 1; return (j);});
+//!
+//! is &int, a reference to a local variable
```

Comment:
TIL 😱

Reply:
> TIL

Wow, I didn't know this either. It's pretty funny. Seems to be C++'s way of providing two functions for the price of one keyword. decltype(variable) gives you declared type of variable name while decltype(expression) gives you type of the expression with an extra & or && added to indicate lvalue or rvalueness (https://stackoverflow.com/questions/27557369/why-does-decltypeauto-return-a-reference-here/27557450#27557450)